language/vm/src/abi.rs

```diff
@@ -0,0 +1,399 @@
+    pub struct_and_function_linking: bool,
+    /// If false, attempting to read structs previously published by this module will fail at runtime
+    pub struct_layout: bool,
+}
+
+impl Compatibility {
+    /// Return true if the two module ABIs compared in the compatiblity check are both linking and
+    /// layout compatible.
+    pub fn is_fully_compatible(&self) -> bool {
+        self.struct_and_function_linking && self.struct_layout
+    }
+}
+
+impl ModuleABI {
+    /// Return compatibility assessment for `new_module` relative to old module `self`
+    pub fn check_compatibility(&self, new_module: &ModuleABI) -> Compatibility {
+        let mut struct_and_function_linking = true;
+        let mut struct_layout = true;
+
+        // module's name and address are unchanged
+        if self.address != new_module.address || self.name != new_module.name {
+            struct_and_function_linking = false;
+        }
+
+        // old module's structs are a subset of the new module's structs
+        for old_struct in &self.structs {
+            match new_module
+                .structs
+                .iter()
+                .find(|s| s.name == old_struct.name)
+            {
+                Some(new_struct) => {
+                    if new_struct.kind != old_struct.kind
+                        || new_struct.type_parameters != old_struct.type_parameters
+                    {
+                        // Declared kind and/or type parameters changed. Existing modules that depend on this struct will fail to link with the new version of the module
+                        struct_and_function_linking = false;
+                        // This does not change the struct layout, but it may leave some published
+                        // values "orphaned". For example: if
+                        // `resource struct S<T: copyable> { t : T }` is changed to
+                        // `resource struct S<T: resource> { t : T}`, code can no longer access
+                        // published values of type (e.g.) `S<u64>`.
+                    }
+                    if new_struct.fields != old_struct.fields {
+                        // Fields changed. Code in this module will fail at runtime if it tries to
+                        // read a previously published struct value
+                        // TODO: this is a stricter definition than required. We could in principle
+                        // choose to label the following as compatible
+                        // (1) changing the name (but not position or type) of a field. The VM does
+                        //     not care about the name of a field (it's purely informational), but
+                        //     clients presumably do.
+                        // (2) changing the type of a field to a different, but layout and kind
+                        //     compatible type. E.g. `struct S { b: bool }` to `struct S { b: B }`
```

Comment:
I don't think this should ever be allowed. Assume you already have a resource of this type stored on-chain. Changing the type will effectively transmute the data.

Reply:
> I don't think this should ever be allowed

I think I agree with you on this philosophically. But it's worth asking whether there are some cases where this is (a) safe from a layout perspective and (b) acceptable from a client perspective. For example: we had a recent bug where `public_key: vector<u8>` allowed an empty vector. It might be possible/desirable for us to allow `public_key: Ed25519PublicKey` where `struct Ed25519PublicKey { bytes: vector<u8> }` and the non-emptiness of `bytes` is enforced at both construction and read time.

To be clear: I think "no" is probably the right policy here, but perhaps worth thinking about the technical feasibility.

Reply:
That's an interesting and legit use case. My take is that if we're making an offline tool then we can go a little bit stricter as the user makes the final decision. However for the long run I'm still thinking about the possibility of having something like this running on-chain, providing some protection for writeset txns. Maybe in that case we could disallow this behavior by default but allow the user to override it explicitly?